### PR TITLE
feat(new): check for installed package managers

### PR DESCRIPTION
**Pull request by onurravli**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Currently, the CLI tool asks for all package managers (npm, yarn, and pnpm) even if they aren't installed.

## What is the new behavior?

Now, CLI checks for installed package managers and then asks for those that are installed, not for the ones that are not installed.

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

### CODE REVIEW

**Comment by onurravli** [comment]
Hey @kamilmysliwiec, can you check this out?

File: actions/new.action.ts [hunk 1]
```diff
@@ -64,6 +64,17 @@ export class NewAction extends AbstractAction {
   }
 }
 
+const isThisPackageManagerInstalled = (packageManager: string) => {
+  try {
+    execSync(`${packageManager} -v`, {
+      stdio: 'ignore',
+    });
+    return true;
+  } catch (error) {
```
Comment (micalevisk) [inline actions/new.action.ts:73]:
```suggestion
  } catch {
```

Comment (onurravli) [reply actions/new.action.ts]: 56493f6d.

**Review by micalevisk — APPROVED** [review]

**Review by Tony133 — CHANGES_REQUESTED** [review]

File: actions/new.action.ts [hunk 2]
```diff
@@ -163,12 +174,17 @@ const installPackages = async (
 };
 
 const askForPackageManager = async (): Promise<Answers> => {
+  const installedPackageManagers = [];
+  if (isThisPackageManagerInstalled('npm'))
+    installedPackageManagers.push(PackageManager.NPM);
+  if (isThisPackageManagerInstalled('yarn'))
+    installedPackageManagers.push(PackageManager.YARN);
+  if (isThisPackageManagerInstalled('pnpm'))
+    installedPackageManagers.push(PackageManager.PNPM);
```
[the remark this answers is not in the document]
Comment (onurravli) [reply actions/new.action.ts]: c517e57.

**Review by Tony133 — APPROVED** [review]

**Comment by kamilmysliwiec** [comment]
Not sure how I feel about this change. Isn't this confusing that some options just won't show up for certain people? Like one might be aware that yarn is a thing but doesn't have it installed on let's say their machine at work. Then they generate a NestJS app and they only see NPM showing up which will make them think using yarn was never an option (simply because it's hidden).

Shouldn't we just leave this decision to the developer since the developer should be aware of what tool they have installed (and even if they don't they can always terminate the command, install it, and start over) and want to use for their projects?

**Comment by Tony133** [comment]
Yes, this is a bit of a "divisive" change, to one 50 percent it can be helpful while the other 50 percent can be confusing 😄

**Comment by onurravli** [comment]
> Not sure how I feel about this change. Isn't this confusing that some options just won't show up for certain people? Like one might be aware that yarn is a thing but doesn't have it installed on let's say their machine at work. Then they generate a NestJS app and they only see NPM showing up which will make them think using yarn was never an option (simply because it's hidden).
> 
> Shouldn't we just leave this decision to the developer since the developer should be aware of what tool they have installed (and even if they don't they can always terminate the command, install it, and start over) and want to use for their projects?

I made this change after I realize there is an error when you continue with yarn even yarn is not installed, as below.

image

If it's not a good idea to hide package managers, it might be a better idea to warn the user if, for example, yarn is not installed, but to build the project (without installing packages).

**Comment by kamilmysliwiec** [comment]
Even if the installation fails, everything else should be properly generated and you should be good to go as long as you manually install dependencies

**Comment by onurravli** [comment]
Yes, I mean the error is not clear. In my firs use of nest-cli, I thought project wasn't built.

**Comment by kamilmysliwiec** [comment]
So maybe instead of hiding package managers we should rather improve our error messages; let users know that even if the installation failed they're generally (almost) all set

**Comment by kamilmysliwiec** [comment]
Feel free to continue in this PR

**Pull request closed without merging**